experimental/list_formatter/src/list_formatter.rs

```diff
@@ -46,86 +39,107 @@ impl ListFormatter {
             2 => {
                 // Pair(values[0], values[1]) = pair_before + values[0] + pair_between + values[1] + pair_after
-                let (before, between, after) = self.data.get().pair(self.width).parts(values[1]);
-                builder = append_literal(builder, before);
-                builder = append_value(builder, values[0]);
-                builder = append_literal(builder, between);
-                builder = append_value(builder, values[1]);
-                append_literal(builder, after)
+                let (before, between, after) = self.data.get().pair(self.width).parts(&values[1]);
+                literal!(before)?;
+                value!(values[0])?;
+                literal!(between)?;
+                value!(values[1])?;
+                literal!(after)
             }
             n => {
                 // Start(values[0], middle(..., middle(values[n-3], End(values[n-2], values[n-1]))...)) =
                 // start_before + values[0] + start_between + (values[1..n-3] + middle_between)* +
                 // values[n-2] + end_between + values[n-1] + end_after
 
                 let (start_before, start_between, _) =
-                    self.data.get().start(self.width).parts(values[1]);
+                    self.data.get().start(self.width).parts(&values[1]);
 
-                builder = append_literal(builder, start_before);
-                builder = append_value(builder, values[0]);
-                builder = append_literal(builder, start_between);
-                builder = append_value(builder, values[1]);
+                literal!(start_before)?;
+                value!(values[0])?;
+                literal!(start_between)?;
+                value!(values[1])?;
 
                 for value in &values[2..n - 1] {
                     let (_, between, _) = self.data.get().middle(self.width).parts(value);
-                    builder = append_literal(builder, between);
-                    builder = append_value(builder, value);
+                    literal!(between)?;
+                    value!(value)?;
                 }
 
                 let (_, end_between, end_after) =
-                    self.data.get().end(self.width).parts(values[n - 1]);
-                builder = append_literal(builder, end_between);
-                builder = append_value(builder, values[n - 1]);
-                append_literal(builder, end_after)
+                    self.data.get().end(self.width).parts(&values[n - 1]);
+                literal!(end_between)?;
+                value!(values[n - 1])?;
+                literal!(end_after)
             }
         }
     }
 
-    fn size_hint(&self, values: &[&str]) -> usize {
-        values.iter().map(|s| s.len()).sum::<usize>()
-            + self.data.get().size_hint(self.width, values.len())
-    }
+    pub fn format<'a, 'b: 'a, 'c: 'a, 'd: 'a, W: FormattedWriteable + 'd>(
+        &'b self,
+        values: &'c [W],
+    ) -> impl FormattedWriteable + 'a {
```

Comment:
Nit: I really do think that we should be returning a concrete type here, like `FormattedList`.  We don't want to be restricted to the methods present on the trait, and we may want to implement additional traits on the concrete type.

Fine to do in a follow-up.